---
title: "Bulk RNA diff gene analysis"
output: 
  html_document:
    self_contained: true
editor_options: 
  chunk_output_type: console
---


## Load R packages
This document depends on the following packages:
```{r Packages, results = "hide", warning = FALSE}

suppressPackageStartupMessages({
  library(biomaRt)
  library(dplyr)
  library(edgeR)
  library(ezRun)
  library(ggfortify)
  library(ggplot2)
  library(Matrix)
  library(matrixStats)
  library(pheatmap)
  library(RColorBrewer)
  library(Rsubread)
  library(scales)
  library(tibble)
  library(tidyverse)
  library(WGCNA)
  library(ggrepel)
  })

```

## TwoGroupsAnalysis_Result {.tabset}

<br/>

### LHFD vs LCHD

#### Load raw count matrix
```{r }
setwd("~/public/_Projects/AT_HFD_Memory/TRAPseq/Cre")
# Load raw RNAseq data
rawCounts <- read.table("~/public/_Sequencing/SEQ00070/aligned/counts/gene_counts_merged.txt", header = TRUE, sep = "\t")
rawCounts <- column_to_rownames(rawCounts, "Geneid")

# Rename samples
colnames(rawCounts) <- gsub("mmepiAT_", "", colnames(rawCounts))
colnames(rawCounts) <- gsub("TRAP.", "", colnames(rawCounts))
colnames(rawCounts) <- gsub("_GRCm38_hisat2", "", colnames(rawCounts))

# Select only HFD and CHD samples
rawCounts <- rawCounts[,grep("CHD|HFD", colnames(rawCounts))]

# Remove genes with no count across all samples
rawCounts <- rawCounts[rowSums(rawCounts) > 0, ]

```

#### Sample metadata
```{r}

# Define sample and conditions
samples   <- colnames(rawCounts)
conditions <- factor(rep(c("CHD", "HFD"), 3))

# Add sample information
metadata <- data.frame(conditions, row.names = samples)

```

#### Normalize raw counts
```{r}

# Create DGEList object
y <- DGEList(counts=rawCounts, group=conditions)

# Remove genes with low expression in > 50% samples
keep <- filterByExpr(y)
y <- y[keep, , keep.lib.sizes=FALSE]

# Apply TMM normalization
y <- calcNormFactors(y)

# Extract normalized counts
normCounts <- cpm(y, log=TRUE)

```

#### Principal component analysis (PCA)
```{r fig.width=12, fig.height=8}

# Run PCA
pcDat  <- prcomp(t(normCounts), scale. = TRUE)

# PC1 vs PC2
autoplot(pcDat, data=metadata,
         colour="conditions", 
         shape=FALSE,
         label.size=3) + theme_classic()

```

#### Sample correlation
```{r fig.width=12, fig.height=8}

colors <- brewer.pal(n = 3, name = "Set2")

# Pearson correlation plot 
pheatmap(
  mat               = cor(normCounts, use="complete.obs"),
  cutree_rows       = 2, 
  cutree_cols       = 2,
  silent            = F,
  annotation_col    = metadata,
  color             = brewer.pal(n = 9, name = "Reds"),
  fontsize_row      = 10, 
  fontsize_col      = 10,
  display_numbers   = FALSE)

```

#### DE analysis

#### Volcano plot
<br/>
Significant DEGs: 2-fold change (log2FC > 1) ; pValue < 0.01
<br/>

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.width=15, fig.height=10}

# Construct design matrix
design <- model.matrix(~ 0 + conditions)
colnames(design) <- levels(conditions)

# Estimate dispersion
y <- estimateDisp(y, design, robust=TRUE)                  

# Estimate quasi-likelihood (QL) dispersions
fitGlm <- glmQLFit(y, design, robust = TRUE)

# Define contrast of interest
contr <- makeContrasts(HFD - CHD, levels=design)

# Apply QL F-test for differential expression
qlf <- glmQLFTest(fitGlm, contrast=contr)

# Extract HFD vs CHD DE results
res <- topTags(qlf, n = Inf, sort.by = "PValue")$table               

# Gene mapping
ensembl <- useMart('ensembl', dataset="mmusculus_gene_ensembl")

geneAnnotation <- getBM(attributes=c("ensembl_gene_id", "external_gene_name"), 
                    filters="ensembl_gene_id", 
                    values=rownames(res), mart=ensembl)

# Combine DE results with gene names
comb <- merge(geneAnnotation, data.frame(res), 
              by.x="ensembl_gene_id", 
              by.y="row.names")

# Rename columns
colnames(comb)[1:2] <- c("geneID", "geneName")

# Annotate genes based on log2FC and FDR thresholds
comb$regulation <- ""
comb[which(comb$logFC > 1 & comb$PValue < 0.01),"regulation"] <- "up"
comb[which(comb$logFC < -1 & comb$PValue < 0.01),"regulation"] <- "down"
comb[which(abs(comb$logFC) <= 1 | comb$PValue >= 0.01),"regulation"] <- "non-significant"

table(comb$regulation)

# Save significant DE genes 
write.table(comb, file = "SEQ070_HFD_vs_CHD_DEGs.txt", row.names = FALSE, col.names = TRUE, sep = "\t", quote = FALSE)

# Remove non-significant genes 
comb <- comb[-grep("^Gm|Rik$", comb$geneName),]

# Show top 10 up and down regulated genes
up_genes   <- subset(comb,regulation=="up")
down_genes <- subset(comb,regulation=="down")

up_genes <- up_genes[order(-up_genes$logFC), ] 
up_genes <- up_genes[1:10,]

down_genes <- down_genes[order(down_genes$logFC), ] 
down_genes <- down_genes[1:10,]

top_genes <- c(up_genes$geneName, down_genes$geneName)
comb$genelabels <- ""

top.idx <- which(comb$geneName %in% c(up_genes$geneName, down_genes$geneName))
comb$genelabels[top.idx] <- comb$geneName[top.idx]

comb$regulation <- factor(comb$regulation)

# Volcano plot
ggplot(comb, aes(logFC, -log10(PValue), color = regulation)) +
  geom_point() + theme_minimal() +
  scale_color_manual(values = c("#00AFBB","#DCDCDC","#FC4E07")) +
  labs(x="log2FC", y="-log10(adjusted pValue)") +
  theme(plot.title = element_text(hjust = 0.5, size = 12, face="bold"), 
        legend.title = element_text(size=10, face="italic")) +
  geom_vline(xintercept=c(-1, 1), lty=2, colour = "#323232") +
  geom_hline(yintercept=-log10(0.01), lty=2, colour = "#323232" ) +
  geom_text_repel(data = comb,
                  aes_string(label = "genelabels"),
                  nudge_x = 0.25, nudge_y = 0.25,
                  max.overlaps = 20,
                  size = 3, segment.size = 0.1, colour="black", 
                  fontface = "bold.italic", show.legend = FALSE) 

```

<br/>


#### Clustering of significant DEGs
<br/>

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.width=15, fig.height=10}

sigDEG <- subset(comb, regulation %in% c("up", "down"))

# Extract normalized counts
lcpm <- cpm(y, log=TRUE)

# Define sample ~ group mapping
mapping <- data.frame(row.names = colnames(rawCounts),
                      conditions)


# Heat map of significant DE genes
pheatmap(lcpm[sigDEG$geneID, ], 
         color=getBlueRedScale(), clustering_method="ward.D2",
         scale="row", cluster_rows=TRUE,
         cluster_cols=TRUE, 
         cutree_rows = 2,
         cutree_cols = 2,
         annotation_col = mapping,
         show_rownames = FALSE,
         show_colnames = TRUE,
         border_color = "black", 
         fontsize_row = 8, fontsize_col = 10,
         annotation_legend = TRUE,
         fontsize=8)
```
<br/>

#### Top up-regulated genes
<br/>

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.width=15, fig.height=10}

# Select top 30 up regulated genes
sigUpGenes <- sigDEG[order(-sigDEG$logFC), ] 
sigUpGenes <- sigUpGenes[1:30,]

# Extract transformed counts for selected genes
upGeneMat <- data.frame(lcpm[sigUpGenes$geneID, ])
upGeneMat <- rownames_to_column(upGeneMat, "geneID")

upGenes <- merge(sigUpGenes, upGeneMat, by.x ="geneID")
upGenes <- upGenes[order(-upGenes$logFC),]

rownames(upGenes) <- NULL
upGenes <- column_to_rownames(upGenes, "geneName")

# Heat map of up regulated genes
pheatmap(upGenes[,9:14], 
         color=getBlueRedScale(), clustering_method="ward.D2",
         scale="row", cluster_rows=FALSE,
         cluster_cols=TRUE, 
         cutree_rows = 2,
         cutree_cols = 2,
         annotation_col = mapping,
         show_rownames = TRUE,
         show_colnames = TRUE,
         border_color = "black", 
         fontsize_row = 10, fontsize_col = 10,
         annotation_legend = TRUE,
         fontsize=8)

```
<br/>

#### Top down-regulated genes
<br/>

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.width=15, fig.height=10}

# Select top 30 up regulated genes
sigDownGenes <- sigDEG[order(sigDEG$logFC), ] 
sigDownGenes <- sigDownGenes[1:30,]

# Extract transformed counts for selected genes
downGeneMat <- data.frame(lcpm[sigDownGenes$geneID, ])
downGeneMat <- rownames_to_column(downGeneMat, "geneID")

downGenes <- merge(sigDownGenes, downGeneMat, by.x ="geneID")
downGenes <- downGenes[order(downGenes$logFC),]

rownames(downGenes) <- NULL
downGenes <- column_to_rownames(downGenes, "geneName")

# Heat map of up regulated genes
pheatmap(downGenes[,9:14], 
         color=getBlueRedScale(), clustering_method="ward.D2",
         scale="row", cluster_rows=FALSE,
         cluster_cols=TRUE, 
         cutree_rows = 2,
         cutree_cols = 2,
         annotation_col = mapping,
         show_rownames = TRUE,
         show_colnames = TRUE,
         border_color = "black", 
         fontsize_row = 10, fontsize_col = 10,
         annotation_legend = TRUE,
         fontsize=8)

```

### HC vs CC

#### Load raw count matrix
```{r }

# Load raw RNAseq data
rawCounts <- read.table("~/public/_Sequencing/SEQ00070/aligned/counts/gene_counts_merged.txt", header = TRUE, sep = "\t")
rawCounts <- column_to_rownames(rawCounts, "Geneid")

# Rename samples
colnames(rawCounts) <- gsub("mmepiAT_", "", colnames(rawCounts))
colnames(rawCounts) <- gsub("TRAP.", "", colnames(rawCounts))
colnames(rawCounts) <- gsub("_GRCm38_hisat2", "", colnames(rawCounts))

# Select only HC and CC samples
rawCounts <- rawCounts[,grep("HC|CC", colnames(rawCounts))]

# Remove genes with no count across all samples
rawCounts <- rawCounts[rowSums(rawCounts) > 0, ]

```

#### Sample metadata
```{r}

# Define sample and conditions
samples   <- colnames(rawCounts)
conditions <- factor(rep(c("CC", "HC"), 3))
# Add sample information
metadata <- data.frame(conditions, row.names = samples)

```

#### Normalize raw counts
```{r}

# Create DGEList object
y <- DGEList(counts=rawCounts, group=conditions)

# Remove genes with low expression in > 50% samples
keep <- filterByExpr(y)
y <- y[keep, , keep.lib.sizes=FALSE]

# Apply TMM normalization
y <- calcNormFactors(y)

# Extract normalized counts
normCounts <- cpm(y, log=TRUE)

```

#### Principal component analysis (PCA)
```{r fig.width=12, fig.height=8}

# Run PCA
pcDat  <- prcomp(t(normCounts), scale. = TRUE)

# PC1 vs PC2
autoplot(pcDat, data=metadata,
         colour="conditions", 
         shape=FALSE,
         label.size=3) + theme_classic()

```

#### Sample correlation
```{r fig.width=12, fig.height=8}

colors <- brewer.pal(n = 3, name = "Set2")

# Pearson correlation plot 
pheatmap(
  mat               = cor(normCounts, use="complete.obs"),
  cutree_rows       = 2, 
  cutree_cols       = 2,
  silent            = F,
  annotation_col    = metadata,
  color             = brewer.pal(n = 9, name = "Reds"),
  fontsize_row      = 10, 
  fontsize_col      = 10,
  display_numbers   = FALSE)

```

#### DE analysis

#### Volcano plot
<br/>
Significant DEGs: 2-fold change (log2FC > 1) ; pValue < 0.01
<br/>

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.width=15, fig.height=10}

# Construct design matrix
design <- model.matrix(~ 0 + conditions)
colnames(design) <- levels(conditions)

# Estimate dispersion
y <- estimateDisp(y, design, robust=TRUE)                  

# Estimate quasi-likelihood (QL) dispersions
fitGlm <- glmQLFit(y, design, robust = TRUE)

# Define contrast of interest
contr <- makeContrasts(HC - CC, levels=design)

# Apply QL F-test for differential expression
qlf <- glmQLFTest(fitGlm, contrast=contr)

# Extract HC vs CC DE results
res <- topTags(qlf, n = Inf, sort.by = "PValue")$table               

# Gene mapping
#ensembl <- useMart('ensembl', dataset="mmusculus_gene_ensembl")

geneAnnotation <- getBM(attributes=c("ensembl_gene_id", "external_gene_name"), 
                    filters="ensembl_gene_id", 
                    values=rownames(res), mart=ensembl)

# Combine DE results with gene names
comb <- merge(geneAnnotation, data.frame(res), 
              by.x="ensembl_gene_id", 
              by.y="row.names")

# Rename columns
colnames(comb)[1:2] <- c("geneID", "geneName")

# Annotate genes based on log2FC and FDR thresholds
comb$regulation <- ""
comb[which(comb$logFC > 1 & comb$PValue < 0.01),"regulation"] <- "up"
comb[which(comb$logFC < -1 & comb$PValue < 0.01),"regulation"] <- "down"
comb[which(abs(comb$logFC) <= 1 | comb$PValue >= 0.01),"regulation"] <- "non-significant"

table(comb$regulation)

# Save significant DE genes 
write.table(comb, file = "SEQ070_HC_vs_CC_DEGs.txt", row.names = FALSE, col.names = TRUE, sep = "\t", quote = FALSE)

# Remove non-significant genes 
comb <- comb[-grep("^Gm|Rik$", comb$geneName),]

# Show top 10 up and down regulated genes
up_genes   <- subset(comb,regulation=="up")
down_genes <- subset(comb,regulation=="down")

up_genes <- up_genes[order(-up_genes$logFC), ] 
up_genes <- up_genes[1:10,]

down_genes <- down_genes[order(down_genes$logFC), ] 
down_genes <- down_genes[1:10,]

top_genes <- c(up_genes$geneName, down_genes$geneName)
comb$genelabels <- ""

top.idx <- which(comb$geneName %in% c(up_genes$geneName, down_genes$geneName))
comb$genelabels[top.idx] <- comb$geneName[top.idx]

comb$regulation <- factor(comb$regulation)

# Volcano plot
ggplot(comb, aes(logFC, -log10(PValue), color = regulation)) +
  geom_point() + theme_minimal() +
  scale_color_manual(values = c("#00AFBB","#DCDCDC","#FC4E07")) +
  labs(x="log2FC", y="-log10(adjusted pValue)") +
  theme(plot.title = element_text(hjust = 0.5, size = 12, face="bold"), 
        legend.title = element_text(size=10, face="italic")) +
  geom_vline(xintercept=c(-1, 1), lty=2, colour = "#323232") +
  geom_hline(yintercept=-log10(0.01), lty=2, colour = "#323232" ) +
  geom_text_repel(data = comb,
                  aes_string(label = "genelabels"),
                  nudge_x = 0.25, nudge_y = 0.25,
                  max.overlaps = 20,
                  size = 3, segment.size = 0.1, colour="black", 
                  fontface = "bold.italic", show.legend = FALSE) 

```
<br/>

#### Clustering of significant DEGs
<br/>

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.width=15, fig.height=10}

sigDEG <- subset(comb, regulation %in% c("up", "down"))

# Extract normalized counts
lcpm <- cpm(y, log=TRUE)

# Define sample ~ group mapping
mapping <- data.frame(row.names = colnames(rawCounts),
                      conditions)


# Heat map of significant DE genes
pheatmap(lcpm[sigDEG$geneID, ], 
         color=getBlueRedScale(), clustering_method="ward.D2",
         scale="row", cluster_rows=TRUE,
         cluster_cols=TRUE, 
         cutree_rows = 2,
         cutree_cols = 2,
         annotation_col = mapping,
         show_rownames = FALSE,
         show_colnames = TRUE,
         border_color = "black", 
         fontsize_row = 8, fontsize_col = 10,
         annotation_legend = TRUE,
         fontsize=8)
```


This document was processed on: `r Sys.Date()`.